bcrypt.js

```diff
@@ -152,15 +152,15 @@ module.exports.compareSync = function compareSync(data, hash) {
 /// @param {String} hash expected hash
 /// @param {Function} cb callback(err, matched) - matched is true if hashed data matches hash
 module.exports.compare = function compare(data, hash, cb) {
-    if (data == null || hash == null) {
+    if (typeof data === 'function') {
         return process.nextTick(function() {
-            cb(new Error('data and hash arguments required'));
+            data(new Error('data and hash arguments required'));
```

Comment:
this doesn't really match the `if` condition. The user might provide a function for the data argument and then see this error when in fact they did provide an argument.

This also doesn't handle the case where the user passes null for the arguments and does not specify a callback. They will not get a rejection.

The real mistake of the API is to not throw then it is called with the wrong arguments which is what should have happened but given that is not the case. Fixing this to match previous behavior means identifying the error condition and then rejecting with promise or callback as needed.